File: src/SparseMatrix.py
class CompressedMatrix:
    def __init__(self, total_rows, total_cols):
        """
        Creates a CompressedMatrix object with specified rows and columns.

        :param total_rows: Number of rows in the matrix.
        :param total_cols: Number of columns in the matrix.
        """
        self.data = {}
        self.total_rows = total_rows
        self.total_cols = total_cols

    @staticmethod
    def read_matrix_file(filepath):
        """
        Reads the matrix content from a file manually (no built-in path libraries).

        :param filepath: Location of the matrix file.
        :return: A list containing the lines read from the file.
        """
        try:
            with open(filepath.replace("\\", "/"), "r") as file:
                content = file.readlines()
                return [line.strip() for line in content if line.strip()]
        except FileNotFoundError:
            raise FileNotFoundError(f"Unable to locate file: {filepath}")

    @staticmethod
    def build_from_file(matrix_filepath):
        """
        Builds a compressed matrix by reading from a given file.

        :param matrix_filepath: The location of the matrix file.
        :return: An instance of CompressedMatrix.
        """
        file_content = CompressedMatrix.read_matrix_file(matrix_filepath)

        if len(file_content) < 2:
            raise ValueError(f"File {matrix_filepath} must contain at least two lines to define dimensions.")
        
        # Extract the matrix dimensions
        try:
            total_rows = int(file_content[0].split('=')[1])
            total_cols = int(file_content[1].split('=')[1])
        except (IndexError, ValueError):
            raise ValueError(f"Invalid dimensions in file {matrix_filepath}.")

        matrix = CompressedMatrix(total_rows, total_cols)

        # Process each matrix element
        for line in file_content[2:]:
            if not line.startswith("(") or not line.endswith(")"):
                raise ValueError(f"Invalid entry format: {line}")

            elements = line[1:-1].split(',')
            try:
                r = int(elements[0].strip())
                c = int(elements[1].strip())
                val = int(elements[2].strip())
            except (IndexError, ValueError):
                raise ValueError(f"Incorrect element format in file: {line}")

            matrix.insert_element(r, c, val)

        return matrix

    def insert_element(self, r, c, val):
        """
        Inserts a value at a specific location in the matrix.

        :param r: Row index.
        :param c: Column index.
        :param val: Value to insert.
        """
        if r >= self.total_rows:
            self.total_rows = r + 1
        if c >= self.total_cols:
            self.total_cols = c + 1

        index = f"{r},{c}"
        self.data[index] = val

    def retrieve_element(self, r, c):
        """
        Retrieves the value at a given matrix position.

        :param r: Row index.
        :param c: Column index.
        :return: The value at the given position or 0 if it doesn't exist.
        """
        index = f"{r},{c}"
        return self.data.get(index, 0)

    def matrix_add(self, other):
        """
        Adds two compressed matrices.

        :param other: Another CompressedMatrix.
        :return: A new CompressedMatrix that is the result of the addition.
        """
        if self.total_rows != other.total_rows or self.total_cols != other.total_cols:
            raise ValueError(f"Matrices must have matching dimensions.")

        result = CompressedMatrix(self.total_rows, self.total_cols)

        # Add elements from the current matrix
        for index, value in self.data.items():
            r, c = map(int, index.split(','))
            result.insert_element(r, c, value)

        # Add elements from the other matrix
        for index, value in other.data.items():
            r, c = map(int, index.split(','))
            result.insert_element(r, c, result.retrieve_element(r, c) + value)

        return result

    def matrix_subtract(self, other):
        """
        Subtracts another matrix from this one.

        :param other: Another CompressedMatrix.
        :return: A new CompressedMatrix representing the result of the subtraction.
        """
        if self.total_rows != other.total_rows or self.total_cols != other.total_cols:
            raise ValueError(f"Matrices must have matching dimensions.")

        result = CompressedMatrix(self.total_rows, self.total_cols)

        # Subtract elements from the other matrix
        for index, value in other.data.items():
            r, c = map(int, index.split(','))
            result.insert_element(r, c, self.retrieve_element(r, c) - value)

        return result

    def matrix_multiply(self, other):
        """
        Multiplies two compressed matrices together.

        :param other: Another CompressedMatrix.
        :return: A new CompressedMatrix representing the product.
        """
        if self.total_cols != other.total_rows:
            raise ValueError(f"Matrix multiplication requires column count of the first to equal row count of the second.")

        result = CompressedMatrix(self.total_rows, other.total_cols)

        for index1, value1 in self.data.items():
            r1, c1 = map(int, index1.split(','))
            for index2, value2 in other.data.items():
                r2, c2 = map(int, index2.split(','))
                if c1 == r2:
                    result.insert_element(r1, c2, result.retrieve_element(r1, c2) + value1 * value2)

        return result

    def __str__(self):
        """
        Provides a string-based view of the compressed matrix.

        :return: A string showing matrix elements and dimensions.
        """
        result = [f"rows={self.total_rows}", f"cols={self.total_cols}"]
        for index, value in self.data.items():
            r, c = index.split(',')
            result.append(f"({r}, {c}, {value})")
        return '\n'.join(result)

    def write_to_file(self, filepath):
        """
        Saves the compressed matrix to a file.

        :param filepath: Destination file path for saving the matrix.
        """
        with open(filepath, 'w') as file:
            file.write(str(self))


def collect_user_input(prompt_text):
    """
    Collects user input from the console.

    :param prompt_text: The message to display to the user.
    :return: Input given by the user.
    """
    return input(prompt_text)


def run_matrix_operations():
    """
    Executes matrix operations based on user commands.
    """
    try:
        available_operations = {
            'A': ('Matrix Addition', 'matrix_add'),
            'B': ('Matrix Subtraction', 'matrix_subtract'),
            'C': ('Matrix Multiplication', 'matrix_multiply')
        }

        print("Available MATRIX Operations: ")
        print("(A) Add Matrices")
        print("(B) Subtract Matrices")
        print("(C) Multiply Matrices")

        operation_choice = collect_user_input("Choose operation (A, B, C): ")
        if operation_choice not in available_operations:
            raise ValueError("Invalid operation selection.")

        filepath1 = collect_user_input("Enter file path for the first matrix: ")
        filepath2 = collect_user_input("Enter file path for the second matrix: ")

        print(f"Loading first matrix from {filepath1}...")
        matrix1 = CompressedMatrix.build_from_file(filepath1)
        print(f"First matrix with size {matrix1.total_rows}x{matrix1.total_cols} loaded successfully.")

        print(f"Loading second matrix from {filepath2}...")
        matrix2 = CompressedMatrix.build_from_file(filepath2)
        print(f"Second matrix with size {matrix2.total_rows}x{matrix2.total_cols} loaded successfully.")

        operation_name, method_name = available_operations[operation_choice]
        print(f"Performing {operation_name}...")

        result_matrix = getattr(matrix1, method_name)(matrix2)
        output_filepath = f"{operation_name}_result.txt"
        result_matrix.write_to_file(output_filepath)

        print(f"Operation completed. Result saved to {output_filepath}.")
    except Exception as error:
        print(f"An error occurred: {str(error)}")


if __name__ == "__main__":
    run_matrix_operations()

File: src/SparseMatrix.js
class SparseMatrixJS {
    constructor(totalRows, totalColumns) {
        /**
         * Initializes the SparseMatrixJS instance.
         *
         * @param {number} totalRows - Total number of rows in the matrix.
         * @param {number} totalColumns - Total number of columns in the matrix.
         */
        this.dataMap = {};
        this.totalRows = totalRows;
        this.totalColumns = totalColumns;
    }

    static readMatrixFromFile(filePath) {
        /**
         * Reads a matrix from a file manually, using basic file input.
         *
         * @param {string} filePath - The path to the matrix file.
         * @returns {Promise<string[]>} A promise resolving to an array of matrix lines.
         */
        return new Promise((resolve, reject) => {
            const fs = require('fs');
            fs.readFile(filePath.replace(/\\/g, "/"), 'utf8', (err, content) => {
                if (err) {
                    reject(`File not found: ${filePath}`);
                } else {
                    const lines = content.split("\n").map(line => line.trim()).filter(line => line);
                    resolve(lines);
                }
            });
        });
    }

    static async fromFile(matrixFilePath) {
        /**
         * Loads a sparse matrix from a file and creates a SparseMatrixJS instance.
         *
         * @param {string} matrixFilePath - The path to the matrix file.
         * @returns {Promise<SparseMatrixJS>} A promise resolving to a SparseMatrixJS instance.
         */
        const fileContent = await SparseMatrixJS.readMatrixFromFile(matrixFilePath);

        if (fileContent.length < 2) {
            throw new Error(`File ${matrixFilePath} must have at least two lines for dimensions.`);
        }

        let totalRows, totalColumns;
        try {
            totalRows = parseInt(fileContent[0].split('=')[1]);
            totalColumns = parseInt(fileContent[1].split('=')[1]);
        } catch (error) {
            throw new Error(`Invalid matrix dimensions in ${matrixFilePath}.`);
        }

        const matrixInstance = new SparseMatrixJS(totalRows, totalColumns);

        fileContent.slice(2).forEach(line => {
            if (!line.startsWith("(") || !line.endsWith(")")) {
                throw new Error(`Invalid element format: ${line}`);
            }
            const [row, col, value] = line.slice(1, -1).split(',').map(val => parseInt(val.trim()));

            matrixInstance.setElement(row, col, value);
        });

        return matrixInstance;
    }

    setElement(row, col, value) {
        /**
         * Inserts a value into the sparse matrix.
         *
         * @param {number} row - Row index for the element.
         * @param {number} col - Column index for the element.
         * @param {number} value - The value to be inserted.
         */
        if (row >= this.totalRows) {
            this.totalRows = row + 1;
        }
        if (col >= this.totalColumns) {
            this.totalColumns = col + 1;
        }
        const key = `${row},${col}`;
        this.dataMap[key] = value;
    }

    getElement(row, col) {
        /**
         * Retrieves an element's value from a specific row and column in the matrix.
         *
         * @param {number} row - Row index.
         * @param {number} col - Column index.
         * @returns {number} Value at the given row and column or 0 if no element is present.
         */
        const key = `${row},${col}`;
        return this.dataMap[key] || 0;
    }

    addMatrices(otherMatrix) {
        /**
         * Adds two sparse matrices together.
         *
         * @param {SparseMatrixJS} otherMatrix - Another sparse matrix.
         * @returns {SparseMatrixJS} New matrix containing the result of the addition.
         */
        if (this.totalRows !== otherMatrix.totalRows || this.totalColumns !== otherMatrix.totalColumns) {
            throw new Error("Matrices have different dimensions.");
        }

        const resultMatrix = new SparseMatrixJS(this.totalRows, this.totalColumns);

        // Add elements from this matrix
        Object.keys(this.dataMap).forEach(key => {
            const [row, col] = key.split(',').map(Number);
            resultMatrix.setElement(row, col, this.dataMap[key]);
        });

        // Add elements from the other matrix
        Object.keys(otherMatrix.dataMap).forEach(key => {
            const [row, col] = key.split(',').map(Number);
            resultMatrix.setElement(row, col, resultMatrix.getElement(row, col) + otherMatrix.dataMap[key]);
        });

        return resultMatrix;
    }

    subtractMatrices(otherMatrix) {
        /**
         * Subtracts another sparse matrix from this one.
         *
         * @param {SparseMatrixJS} otherMatrix - Another sparse matrix.
         * @returns {SparseMatrixJS} New matrix representing the result of the subtraction.
         */
        if (this.totalRows !== otherMatrix.totalRows || this.totalColumns !== otherMatrix.totalColumns) {
            throw new Error("Matrices have different dimensions.");
        }

        const resultMatrix = new SparseMatrixJS(this.totalRows, this.totalColumns);

        Object.keys(otherMatrix.dataMap).forEach(key => {
            const [row, col] = key.split(',').map(Number);
            resultMatrix.setElement(row, col, this.getElement(row, col) - otherMatrix.dataMap[key]);
        });

        return resultMatrix;
    }

    multiplyMatrices(otherMatrix) {
        /**
         * Multiplies two sparse matrices together.
         *
         * @param {SparseMatrixJS} otherMatrix - Another sparse matrix.
         * @returns {SparseMatrixJS} Result matrix after multiplication.
         */
        if (this.totalColumns !== otherMatrix.totalRows) {
            throw new Error("Matrix multiplication is only possible when the number of columns in the first matrix equals the number of rows in the second matrix.");
        }

        const resultMatrix = new SparseMatrixJS(this.totalRows, otherMatrix.totalColumns);

        Object.keys(this.dataMap).forEach(key1 => {
            const [row1, col1] = key1.split(',').map(Number);
            Object.keys(otherMatrix.dataMap).forEach(key2 => {
                const [row2, col2] = key2.split(',').map(Number);
                if (col1 === row2) {
                    resultMatrix.setElement(row1, col2, resultMatrix.getElement(row1, col2) + this.dataMap[key1] * otherMatrix.dataMap[key2]);
                }
            });
        });

        return resultMatrix;
    }

    toString() {
        /**
         * Converts the sparse matrix to a string representation.
         *
         * @returns {string} Stringified representation of the matrix and its elements.
         */
        const matrixLines = [`rows=${this.totalRows}`, `cols=${this.totalColumns}`];
        Object.keys(this.dataMap).forEach(key => {
            const [row, col] = key.split(',');
            matrixLines.push(`(${row}, ${col}, ${this.dataMap[key]})`);
        });
        return matrixLines.join('\n');
    }

    async saveMatrixToFile(filePath) {
        /**
         * Saves the sparse matrix to a file.
         *
         * @param {string} filePath - The file path where the matrix will be saved.
         */
        const fs = require('fs');
        fs.writeFileSync(filePath, this.toString(), 'utf8');
    }
}

function getUserInput(promptText) {
    /**
     * Collects user input from the terminal.
     *
     * @param {string} promptText - Text displayed as a prompt to the user.
     * @returns {Promise<string>} A promise resolving to the user input.
     */
    const readline = require('readline');
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    return new Promise((resolve) => {
        rl.question(promptText, (input) => {
            rl.close();
            resolve(input);
        });
    });
}

async function executeMatrixOperations() {
    /**
     * Handles matrix operations based on user input.
     */
    try {
        const operations = {
            'A': { operationName: 'Matrix Addition', method: 'addMatrices' },
            'B': { operationName: 'Matrix Subtraction', method: 'subtractMatrices' },
            'C': { operationName: 'Matrix Multiplication', method: 'multiplyMatrices' }
        };

        console.log("Available MATRIX Operations:");
        console.log("(A) Addition");
        console.log("(B) Subtraction");
        console.log("(C) Multiplication");

        const operationChoice = await getUserInput("Choose an operation (A, B, C): ");
        if (!operations[operationChoice]) {
            throw new Error("Invalid option selected.");
        }

        const matrixFile1 = await getUserInput("Enter path for the first matrix file: ");
        const matrixFile2 = await getUserInput("Enter path for the second matrix file: ");

        console.log(`Loading first matrix from ${matrixFile1}...`);
        const matrix1 = await SparseMatrixJS.fromFile(matrixFile1);
        console.log(`First matrix loaded with dimensions ${matrix1.totalRows}x${matrix1.totalColumns}`);

        console.log(`Loading second matrix from ${matrixFile2}...`);
        const matrix2 = await SparseMatrixJS.fromFile(matrixFile2);
        console.log(`Second matrix loaded with dimensions ${matrix2.totalRows}x${matrix2.totalColumns}`);

        const selectedOperation = operations[operationChoice];
        const resultMatrix = matrix1[selectedOperation.method](matrix2);
        console.log(`${selectedOperation.operationName} Result:`);
        console.log(resultMatrix.toString());

        const saveFilePath = await getUserInput("Enter path to save the result matrix: ");
        await resultMatrix.saveMatrixToFile(saveFilePath);
        console.log(`Matrix saved successfully to ${saveFilePath}`);

    } catch (error) {
        console.error(`Error: ${error.message}`);
    }
}

executeMatrixOperations();
